ops/model.py

```diff
@@ -1559,6 +1560,14 @@ def _run(self, *args, return_output=False, use_json=False):
     def _is_relation_not_found(model_error):
         return 'relation not found' in str(model_error)
 
+    def check_relation_validity(self, relation_name):
```

Comment:
Regarding your first point - the entire backend class is private - and shouldn't be used directly by users.  So a private method on a private class just seemed a bit redundant.  For your second point, we aren't simply checking if a peer relation doesn't exist at harness-code execution time.  We are trying to intercept relation data access (within the charm code).  I agree that keeping the concerns separate is valuable.  Strictly speaking, this doesn't directly tangle those concerns because the model code just calls a generic `check` function on the backend and the backends are responsible for deciding what is appropriate.  But I agree it doesn't feel great.